Sources/AsyncHTTPClient/ConnectionPool/ChannelHandler/SOCKSEventsHandler.swift

```diff
@@ -0,0 +1,117 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOSOCKS
+
+final class SOCKSEventsHandler: ChannelInboundHandler, RemovableChannelHandler {
+    typealias InboundIn = NIOAny
+
+    enum State {
+        // transitions to channelActive or failed
+        case initialized
+        // transitions to socksEstablished or failed
+        case channelActive(Scheduled<Void>)
+        // final success state
+        case socksEstablished
+        // final success state
+        case failed(Error)
+    }
+
+    private var socksEstablishedPromise: EventLoopPromise<Void>?
+    var socksEstablishedFuture: EventLoopFuture<Void>! {
+        return self.socksEstablishedPromise?.futureResult
+    }
+
+    private let deadline: NIODeadline
+    private var state: State = .initialized
+
+    init(deadline: NIODeadline) {
+        self.deadline = deadline
+    }
+
+    func handlerAdded(context: ChannelHandlerContext) {
+        self.socksEstablishedPromise = context.eventLoop.makePromise(of: Void.self)
+
+        if context.channel.isActive {
+            self.connectionStarted(context: context)
+        }
+    }
+
+    func handlerRemoved(context: ChannelHandlerContext) {
+        struct NoResult: Error {}
+        self.socksEstablishedPromise!.fail(NoResult())
```

Comment:
How can we make sure that the right error is thrown if the error is private to a function?

Reply:
This error is only really used, if a developer removes the Handler from the pipeline before a success or failure has happened (This might happen in HappyEyeballs for example). Promises must be fulfilled, before they can be released (otherwise crash!). To ensure we don't crash, this trick is used here. This is part of the dirty tricks in the transport handshake state to somewhere outside the pipeline. I'm not sure if we want to test for this error. Since this is all internal API only, I would think we can leave it as is.